### PR TITLE
Allow overwriting members of a WorkspaceGroup when setting output properties to the same workspace name

### LINKED ISSUE #12717: Running an algorithm on a workspace group with input and output set to the same name does not behave as expected
Since with any other workspace type this would overwrite the original workspace the same should be done with groups, i.e. the contents of the group are overwritten with the results.

 - - - - 
Keywords: core

### PR DESCRIPTION
**Pull request by DanNixon**
Fixes #12717 

This adds the ability for an algorithm to overwrite the members of a WorkspaceGroup when processing workspace groups when an output property is given the same name as an input property, before this would have created new workspaces with `name_n` style names, which is inconsistent with what you would expect given how algorithms run over single workspaces.

To test:
- Code review
- Try some algorithms that take multiple input and output workspaces and see that it behaves as expected

### CODE REVIEW

File: Code/Mantid/Framework/API/src/Algorithm.cpp [hunk 1]
```diff
@@ -1324,8 +1338,24 @@ bool Algorithm::processGroups() {
...
+
+        // Overwrite workspaces in any input property if they have the same
+        // name as an output (i.e. copy name button in algorithm dialog used)
+        // (only need to do this for a single input, multiple will be handled
+        // by ADS)
+        if (inputProperty && inputProperty->value() == prop->value())
+          outName = m_groups[inputProp - m_inputWorkspaceProps.begin()][entry]
+                        ->name();
+
         // Except if all inputs had similar names, then the name is "out_1"
```
Comment (mantid-roman) [inline Code/Mantid/Framework/API/src/Algorithm.cpp]: Is it necessary to create inputProp variable and do the second if? It looks to me that the second if's condition is always true when the first if is true.

Comment (DanNixon) [inline Code/Mantid/Framework/API/src/Algorithm.cpp]: @mantid-roman It protects against the ```dynamic_cast``` failing, although there shouldn't really be a case where it does.

Comment (mantid-roman) [inline Code/Mantid/Framework/API/src/Algorithm.cpp]: I meant inputProperty. It is only used to repeat the checks done in WorkspacePropertyValueIs.

Comment (DanNixon) [inline Code/Mantid/Framework/API/src/Algorithm.cpp]: I see what you were meaning now, I could remove the first ```if```.

File: Code/Mantid/Framework/API/src/Algorithm.cpp [hunk 2]
```diff
@@ -1324,8 +1338,22 @@ bool Algorithm::processGroups() {
...
+        // Overwrite workspaces in any input property if they have the same
+        // name as an output (i.e. copy name button in algorithm dialog used)
+        // (only need to do this for a single input, multiple will be handled
+        // by ADS)
+        if (inputProp != m_inputWorkspaceProps.end() &&
+            dynamic_cast<Property *>(*inputProp)->value() == prop->value())
+          outName = m_groups[inputProp - m_inputWorkspaceProps.begin()][entry]
+                        ->name();
+
         // Except if all inputs had similar names, then the name is "out_1"
```
Comment (mantid-roman) [inline Code/Mantid/Framework/API/src/Algorithm.cpp]: Wouldn't ```if (inputProp != m_inputWorkspaceProps.end())``` be enough here?

Comment (DanNixon) [inline Code/Mantid/Framework/API/src/Algorithm.cpp]: Yes. I have no idea what I was previously thinking.

File: Code/Mantid/Framework/API/src/Algorithm.cpp [hunk 3]
```diff
@@ -39,6 +39,20 @@ namespace API {
 namespace {
 /// Separator for workspace types in workspaceMethodOnTypes member
 const std::string WORKSPACE_TYPES_SEPARATOR = ";";
+
+class WorkspacePropertyValueIs {
+public:
+  WorkspacePropertyValueIs(std::string value) : m_value(value){};
```
Comment (mantid-roman) [inline Code/Mantid/Framework/API/src/Algorithm.cpp]: Sorry I didn't see it earlier but ```value``` should be passed by const reference.

**Pull request merged**